libcaf_core/caf/flow/op/buffer.test.cpp

```diff
@@ -0,0 +1,499 @@
+                                       snk->as_observer());
+        add_subs(uut);
+        snk->request(42);
+        run_flows();
+        uut->fwd_on_next(fwd_ctrl, 1);
+        run_flows();
+        check(snk->buf.empty());
+      }
+    }
+    WHEN("the control observable fires with pending data") {
+      THEN("the operator emits a partial batch") {
+        auto snk = flow::make_passive_observer<cow_vector<int>>();
+        auto grd = make_unsubscribe_guard(snk);
+        auto uut = raw_sub<skip_trait>(3, trivial_obs<int>(),
+                                       trivial_obs<int64_t>(),
+                                       snk->as_observer());
+        add_subs(uut);
+        snk->request(42);
+        run_flows();
+        uut->fwd_on_next(fwd_data, 17);
+        uut->fwd_on_next(fwd_ctrl, 1);
+        run_flows();
+        check_eq(snk->buf, std::vector{cow_vector<int>{17}});
+      }
+    }
+  }
+}
+
+SCENARIO("no-skip policies emit empty batches") {
+  GIVEN("a buffer operator") {
+    WHEN("the control observable fires with no pending data") {
+      THEN("the operator emits an empty batch") {
+        auto snk = flow::make_passive_observer<cow_vector<int>>();
+        auto grd = make_unsubscribe_guard(snk);
+        auto uut = raw_sub<noskip_trait>(3, trivial_obs<int>(),
+                                         trivial_obs<int64_t>(),
+                                         snk->as_observer());
+        add_subs(uut);
+        snk->request(42);
+        run_flows();
+        uut->fwd_on_next(fwd_ctrl, 1);
+        run_flows();
+        check_eq(snk->buf, std::vector{cow_vector<int>()});
+      }
+    }
+    WHEN("the control observable fires with pending data") {
+      THEN("the operator emits a partial batch") {
+        auto snk = flow::make_passive_observer<cow_vector<int>>();
+        auto grd = make_unsubscribe_guard(snk);
+        auto uut = raw_sub<noskip_trait>(3, trivial_obs<int>(),
+                                         trivial_obs<int64_t>(),
+                                         snk->as_observer());
+        add_subs(uut);
+        snk->request(42);
+        run_flows();
+        uut->fwd_on_next(fwd_data, 17);
+        uut->fwd_on_next(fwd_ctrl, 1);
+        run_flows();
+        check_eq(snk->buf, std::vector{cow_vector<int>{17}});
+      }
+    }
+  }
+}
+
+SCENARIO("disposing a buffer operator completes the flow") {
+  GIVEN("a buffer operator") {
+    WHEN("disposing the subscription operator of the operator") {
+      THEN("the observer receives an on_complete event") {
+        auto snk = flow::make_passive_observer<cow_vector<int>>();
+        auto uut = raw_sub<skip_trait>(3, trivial_obs<int>(),
+                                       trivial_obs<int64_t>(),
+                                       snk->as_observer());
+        add_subs(uut);
+        snk->request(42);
+        run_flows();
+        uut->dispose();
+        run_flows();
```

Comment:
I think this is the better (and safer) option to clean up tests. Please drop the unsubscribe guard and use this instead. Can be a convenience function in the fixture or just add these two lines after each test setup.

With this gone as well, I believe the fixture should remain unchanged. Which is a good thing, we had quite a bit of unnecessary complexity in there that we should actively strive to get rid of.

Reply:
We still have `passive_subscription_impl` that was added to this fixture. Should I look into removing it from the fixture as well?

Reply:
Yep, give it a shot. Less is more.